posts/2021-11-26-jakarta-ee-9.1.adoc

```diff
@@ -0,0 +1,430 @@
+
+* link:https://github.com/OpenLiberty/open-liberty/issues/17599[HTTP upgrade to WebSocket can cause quiesce errors]
++
+When a websocket connection is started, it starts as an HTTP connection. Previously, If an error occured during the transition between an HTTP and a websocket connection, which was known to be a websocket upgrade, the error processing would neglect to decrement a connection counter, which then caused the server to believe there is an open connection during server stop.
+There were two scenarios where these quiesce errors would occur:
++
+** When a read error occured during the transition between an HTTP and a websocket connection, the error processing neglected to decrement a connection counter, which then causes the server to believe there is an open connection during server stop.
++
+** If a client immediately closed the websocket connection after it was opened, the original upgrade request handling may not have had enough time to close properly on the server. Once again, the connection counter failed to decrement leading the server to believe there is an open connection during the server stop.
++
+This issue has been fixed by adding a new flag called `decrementNeeded` which helps to ensure that the decrement is not neglected.
+
+
+* link:https://github.com/OpenLiberty/open-liberty/issues/19227[Ensure ServletRequestListener#requestDestroyed is always called]
++
+We discovered a bug where the `ServletRequestListener#requestDestroyed` call does not occur, if an exception occurs during async servlet while an `appSecurity-x.0` is enabled. For this bug to occur, two conditions must be met: the `webContainer` property `deferServletRequestListenerDestroyOnError` is true and an `appSecurity-x.0` feature is enabled. This issue has now been resolved.
+
+* link:https://github.com/OpenLiberty/open-liberty/issues/19197[ClassCastException in JSP relating to JDT internal classes]
++
+Open Liberty 21.0.0.9 introduced a bug where the following error occured for certain class lookups in JSP:
++
+[source]
+----
+Error 500: java.lang.ClassCastException: class org.eclipse.jdt.internal.compiler.lookup.PlainPackageBinding cannot be cast to class org.eclipse.jdt.internal.compiler.lookup.TypeBinding (org.eclipse.jdt.internal.compiler.lookup.PlainPackageBinding and org.eclipse.jdt.internal.compiler.lookup.TypeBinding are in unnamed module of loader org.eclipse.osgi.internal.loader.EquinoxClassLoader @3522bc53)
+----
++
+This issue has now been fixed, as the private `isPackage(String)` method needed to be updated to correctly identify that `org.apache.jsp.test_jsp$1InnerClass` is a class and not a package.
```

Comment:
I think it's better to leave this out, and only say "This issue has now been fixed".

Reply:
Actually, you can leave it. But change the example to `com.ibm._jsp._test_JSP$1InnerClass` instead.

Reply:
```suggestion
This issue has now been fixed.
```